File: qt/Meteorienteer/taivaanvahtifield.cpp
#include "taivaanvahtifield.h"
#include <QDebug>
#include <QDomNode>

TaivaanvahtiField::TaivaanvahtiField(QObject *parent) :
    QObject(parent), mandatory(false), fieldType(TYPE_NOT_SET)
{
}

void TaivaanvahtiField::parseFieldElement(QDomElement elem)
{
    fieldId = elem.firstChildElement("field_id").text();
    fieldLabel = elem.firstChildElement("field_label").text();
    mandatory = elem.firstChildElement("field_mandatory").text() == "1";
    fieldInfo = elem.firstChildElement("field_info").text();
    fieldInfoUrl = QUrl(elem.firstChildElement("field_info_url").text());

    QString fieldTypeString = elem.firstChildElement("field_type").text();

    qDebug() << "parsing field id:" << fieldId << "field type string:" << fieldTypeString;

    if(fieldTypeString=="text") fieldType = TYPE_TEXT;
    if(fieldTypeString.toLower()=="date") fieldType = TYPE_DATE;
    if(fieldTypeString.toLower()=="time") fieldType = TYPE_TIME;
    if(fieldTypeString=="coordinate") fieldType = TYPE_COORDINATE;
    if(fieldTypeString=="checkbox") fieldType = TYPE_CHECKBOX;
    if(fieldTypeString=="select") fieldType = TYPE_SELECTION;
    if(fieldTypeString=="specific_tulipallon_kirkkaus") fieldType = TYPE_SELECTION;
    if(fieldType == TYPE_NOT_SET) {
        qDebug() << Q_FUNC_INFO << "Warning: field type " << fieldTypeString << " not handled!";
    }
    qDebug() << "evaluated as:" << fieldType;
    QDomElement valuesElement = elem.firstChildElement("values");
    if(!valuesElement.isNull()) {
        QDomElement valueElement = valuesElement.firstChildElement("value");
        while(!valueElement.isNull()) {
            fieldValues.insert(valueElement.firstChildElement("value_id").text(), valueElement.firstChildElement("value_name").text());
            valueElement = valueElement.nextSiblingElement("value");
        }
    }
    if(!fieldValues.isEmpty()) fieldType = TYPE_SELECTION;
    // qDebug() << fieldId << fieldLabel << fieldValues;
}

void TaivaanvahtiField::createFieldElement(QDomElement elem, QString value, QDomDocument &doc)
{
    QString idString = id();
    // Workarounds:
    if(idString == "observation_start_hours") idString = "start_hours";
    if(idString == "observation_end_hours") idString = "end_hours";
    createFieldElement(elem, idString, value, doc);
}

void TaivaanvahtiField::createFieldElement(QDomElement elem, QString name, QString value, QDomDocument &doc)
{
    QDomElement fieldElement = doc.createElement("field");
    elem.appendChild(fieldElement);

    QDomElement fieldIdElement = doc.createElement("field_id");
    fieldElement.appendChild(fieldIdElement);

    QDomText fieldIdText = doc.createTextNode(name);
    fieldIdElement.appendChild(fieldIdText);

    QDomElement fieldValueElement = doc.createElement("field_value");
    fieldElement.appendChild(fieldValueElement);
    QDomText fieldValueText = doc.createTextNode(value);
    fieldValueElement.appendChild(fieldValueText);
}

File: MeteorienteerU/Assets/Scripts/TaivaanvahtiField.cs
using UnityEngine;
using System.Collections.Generic;
using System.Xml;

public class TaivaanvahtiField
{
 	public enum FieldType {
        TYPE_NOT_SET,
        TYPE_TEXT,
        TYPE_CHECKBOX,
        TYPE_SELECTION,
        TYPE_DATE,
        TYPE_COORDINATE,
        TYPE_TIME
    };
	
	public string id {get; set;}
	public string label {get; private set;}
	public string info {get; private set;}
	public string infoUrl {get; private set;}
	public bool isMandatory {get; private set;}
	public FieldType type {get; private set;}
	public Dictionary<string, string> values {get; private set;}
		
	public TaivaanvahtiField(TaivaanvahtiForm form)
	{
	}
	
	public void ParseFieldElement(XmlNode node)
	{
		id = node["field_id"].InnerText;
	    label = node["field_label"].InnerText;
		isMandatory = node["field_mandatory"].InnerText == "1";
		info = node["field_info"].InnerText;
		infoUrl = node["field_info_url"].InnerText;
		
		type = FieldType.TYPE_NOT_SET;
	    string fieldTypeString = node["field_type"].InnerText;
	    if(fieldTypeString=="text") type = FieldType.TYPE_TEXT;
	    if(fieldTypeString.ToLower()=="date") type = FieldType.TYPE_DATE;
	    if(fieldTypeString.ToLower()=="time") type = FieldType.TYPE_TIME;
	    if(fieldTypeString=="coordinate") type = FieldType.TYPE_COORDINATE;
	    if(fieldTypeString=="checkbox") type = FieldType.TYPE_CHECKBOX;
	    if(fieldTypeString=="select") type = FieldType.TYPE_SELECTION;
	    if(fieldTypeString=="specific_tulipallon_kirkkaus") type = FieldType.TYPE_SELECTION;
	    if(type == FieldType.TYPE_NOT_SET) {
       	 Debug.Log("Warning: field type " + fieldTypeString + " not handled!");
    	}
		
		XmlNode valuesNode = node["values"];
		if (valuesNode != null)
		{
			values = new Dictionary<string, string>();
			XmlNode valueNode = valuesNode["value"];
			while (valueNode != null)
			{
				values.Add(valueNode["value_id"].InnerText, valueNode["value_name"].InnerText);
				do
				{
					valueNode = valueNode.NextSibling;
				}
				while (valueNode != null && valueNode.Name != "value");
			}
		}
		
		if (values != null && values.Count > 0) type = FieldType.TYPE_SELECTION;
	}
	
    public void CreateFieldElement(XmlElement elem, string valuestr, XmlDocument doc)
	{
	    string idString = id;
    	// Workarounds:
	    if(idString == "observation_start_hours") idString = "start_hours";
	    if(idString == "observation_end_hours") idString = "end_hours";
	    CreateFieldElement(elem, idString, valuestr, doc);
	}
    
	public static void CreateFieldElement(XmlElement elem, string name, string valuestr, XmlDocument doc)
	{
	    XmlElement fieldElement = doc.CreateElement("field");
	    elem.AppendChild(fieldElement);
	
	    XmlElement fieldIdElement = doc.CreateElement("field_id");
	    fieldElement.AppendChild(fieldIdElement);
	
	    XmlText fieldIdText = doc.CreateTextNode(name);
	    fieldIdElement.AppendChild(fieldIdText);
	
	    XmlElement fieldValueElement = doc.CreateElement("field_value");
	    fieldElement.AppendChild(fieldValueElement);
	    XmlText fieldValueText = doc.CreateTextNode(valuestr);
	    fieldValueElement.AppendChild(fieldValueText);
	}
}
